### ISSUE TITLE
Django Channels Memory Leak on every message or connection

### ISSUE THREAD

**Issue by cacosandon** [django/channels#2094]
I'm having a memory leak in Django Channels using `uvicorn`.

Every "memory crash" is a restart/deploy.
<img width="450" alt="Screenshot 2024-04-30 at 08 44 06" src="https://github.com/django/channels/assets/30879716/c68b839f-a877-4856-b08d-ad6af75e6549">

I initially thought that could be my fault, but I've followed the tutorial and it happens also with that.

Here is the repository with that minimal example and memory profiling: https://github.com/cacosandon/django-channels-memory-leak

This happens locally, in the server, with/without `DEBUG`, just by reconnecting or sending messages (in the example I've added large messages so you can notice the memory leak).

The memory is never released.. even if the user disconnects after. 
<img width="700" alt="Screenshot 2024-04-30 at 08 55 10" src="https://github.com/django/channels/assets/30879716/74d2ce97-e86b-48d5-9d7e-7f2f95090593">

I've proved it with `memory-profiler` and `memray` (both commands were added in the README so you can reproduce)

Dependencies:

```
Django==5.0.4
channels==4.0.0
channels-redis==4.2.0
uvicorn[standard]==0.29.0

# Profiling memory
memory-profiler==0.61.0
memray==1.12.0
``` 

I (think that) have really tried everything; deleting objects, manual garbage collection, etc. Nothing prevents the memory to increase and to never be released back. Any insights? 🙏

**Comment by cacosandon** [comment]
I've tested `daphne` and `hypercorn` alongside `uvicorn`. All three show a similar pattern of memory usage, increasing steadily up to around 160 MiB. Despite this, they continue to consume more memory indefinitely, as monitored by `memory-profiler`.

The interesting thing is, while `uvicorn` shows a continuous rise in memory usage on the `memray` graph, the graphs for `daphne` and `hypercorn` are flat at 80 MiB. This discrepancy makes it unclear which tool provides more reliable data.

Here are the commands I used for each:

- Uvicorn: `memray run --force -o output.bin -m uvicorn core.asgi:application`
- Daphne: `python -m memray run -o output.bin --force ./manage.py runserver`
- Hypercorn: `memray run --force -o output.bin -m hypercorn core.asgi:application`

**Comment by carltongibson** [comment]
And can you see from any of the tools, memray perhaps, which objects are consuming the memory?

(I'd expect a `gc.collect()` to help here TBH)

**Comment by carltongibson** [comment]
@cacosandon Also, can you try with the PubSub layer, and see if the results are different there? Thanks.

**Comment by cacosandon** [comment]
Sure! I'll try to find time today to prepare a report on `memray --leaks` for each protocol server and test the PubSub layer. I'll get back to you soon, thanks!

**Comment by cacosandon** [comment]
So I tried multiple combinations. All HTML reports from memray are here: 
[reports.zip](https://github.com/django/channels/files/15178715/reports.zip)

But below there are screenshots from them.

First, tried with Redis Channels (not PubSub) to get memory leaks.

With `uvicorn`

`PYTHONMALLOC=malloc memray run --force -o output.bin -m uvicorn core.asgi:application`
+
`memray flamegraph output.bin --force --leaks`


So, the leaks report include memory that was never released back, but I don't know how to interpret it correctly. Seems like `AuthMiddleware` was leaking but after removing it, the results are almost the same.

`redis-channels-uvicorn-leaks.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 46 10" src="https://github.com/django/channels/assets/30879716/36181744-41c2-4129-96bc-6f670ad97080">


Here is the screenshot of the `uvicorn` + leaks without `AuthMiddleware`:

`redis-channels-uvicorn-without-authmiddleware-leaks.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 52 05" src="https://github.com/django/channels/assets/30879716/a57acc7b-dda8-474b-a4f5-ea6b13c9ef83">

Then tried with `daphne`

`PYTHONMALLOC=malloc memray run --force -o output.bin -m daphne core.asgi:application`
+
`memray flamegraph output.bin --force --leaks`

`redis-channels-daphne-leaks.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 50 50" src="https://github.com/django/channels/assets/30879716/e5542d76-a0e4-429f-9016-791306bd42cd">


The interesting part is that `hypercorn` showed no memory leaks (or maybe `memray` is not working here(?))

`PYTHONMALLOC=malloc memray run --force -o output.bin -m hypercorn core.asgi:application`
+
`memray flamegraph output.bin --force --leaks`

`redis-channels-hypercorn-leaks.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 54 14" src="https://github.com/django/channels/assets/30879716/f5bd60ac-189a-4f2d-917a-91d349cddace">



-------

Then, I tried with garbage collect for `uvicorn` and `daphne`. Same story for both.


`memray run --force -o output.bin -m uvicorn core.asgi:application`
+
`memray flamegraph output.bin --force`

`redis-channels-uvicorn-gccollect.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 55 31" src="https://github.com/django/channels/assets/30879716/9f25c046-76ed-4561-957f-9ff0420f44d8">

`memray run --force -o output.bin -m daphne core.asgi:application`
+
`memray flamegraph output.bin --force`

`redis-channels-daphne-gccollect.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 56 03" src="https://github.com/django/channels/assets/30879716/45e52668-cc29-4b1b-898a-22fbc2350ba5">


-------

And finally tried with `PubSub` for `uvicorn` and `daphne`

`memray run --force -o output.bin -m uvicorn core.asgi:application`
+
`memray flamegraph output.bin --force`

`redis-pubsub-uvicorn.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 57 34" src="https://github.com/django/channels/assets/30879716/fc75d492-47dd-480c-aa57-2c165729ad5d">


`memray run --force -o output.bin -m daphne core.asgi:application`
+
`memray flamegraph output.bin --force`

`redis-pubsub-daphne.html`
<img width="600" alt="Screenshot 2024-05-01 at 08 57 56" src="https://github.com/django/channels/assets/30879716/12428afe-af9a-4fdd-be2a-093235f1bc84">


------

Just in case, I also removed all `@profile` above functions so the memory leaks were not affected by the `memory-profiler` library.

Hope all these reports help understanding the constant memory increase.

Right now I am trying to move my application to `hypercorn` so I can test it on staging, but some group messages are not being sent. If I manage to solve it, I'll post the results here!

**Comment by cacosandon** [comment]
I've made it to make `hypercorn` work!

For some reason the websocket messages that were bytes-only were sent as `{"text": None, "bytes": ... }` so the function of `AsyncWebsocketConsumer` always called the text handler. 

Will report it in another issue later!

```diff
  async def websocket_receive(self, message):
      """
      Called when a WebSocket frame is received. Decodes it and passes it
      to receive().
      """
-     if "text" in message:
+     if "text" in message and message["text"] is not None:
          await self.receive(text_data=message["text"])
      else:
          await self.receive(bytes_data=message["bytes"])
```

Testing now in staging 🤞

**Comment by cacosandon** [comment]
Ok, seems that with `hypercorn` the memory leak is fixed! 🥳 

<img width="608" alt="Screenshot 2024-05-01 at 18 26 34" src="https://github.com/django/channels/assets/30879716/298da3fe-ebeb-4cde-91e1-ebef23619344">

But `daphne` and `uvicorn` are causing problems. Is there anything I can help to figure out why?

**Comment by cacosandon** [comment]
Seems that still there is a memory leak in my application 😕 

<img width="598" alt="Screenshot 2024-05-01 at 22 40 21" src="https://github.com/django/channels/assets/30879716/9aac9cce-bbd8-45a3-ae03-bad268c42cfd">

**Comment by carltongibson** [comment]
Hi @cacosandon 

Looking at the uploaded report, for e.g. `redis-pubsub-daphe`, the memory usage rises and the stabilises: 

![Screenshot 2024-05-02 at 08 21 09](https://github.com/django/channels/assets/64686/8134b81d-5567-412c-9f4f-624bc2f33d2c)

**Comment by cacosandon** [comment]
@carltongibson, do you have any clue about what's happening? Or what else can I try? I'm willing to try anything!

**Comment by carltongibson** [comment]
@cacosandon Given that you report it happening with the pub sub layer and different servers, not really. You need to identify where the leak is happening. Then it's possible to say something.

**Comment by cacosandon** [comment]
@carltongibson all my samples are from using `RedisChannelLayer` or `RedisPubSubLayer`, with `uvicorn`, `daphne` or `hypercorn`, with the tutorial example. My app has the problem too but I think it's a generalized problem.

Some things I've noticed:
- Memory increases constantly when there are **large** messages (>0.5MiB)
- Memory increases constantly when there are multiple connects/disconnects (every handshake adds memory)
- Memory leak **is not present** using `InMemoryChannelLayer`
- Using explicit `del` and `gc.collect()` decelerates the increase of memory.. but the leak is still present
- Creating large objects in Django Views does not leak the memory (every request kinds of clean up)

I don't know how nobody else is having this problem. Maybe they just don't send large messages 🤔

**Comment by carltongibson** [comment]
Hi @cacosandon — are you able to identify where the leak is happening? Alas, I haven't had time to dig into this further for you.  Without that it's difficult to say too much. 

If you can identify a more concrete issue, there's a good chance we can resolve it.

**Comment by cacosandon** [comment]
@carltongibson no :( that's actually the thing that I'm struggling on: finding the memory leak 😓 

I really tried every tool to detect it, but nothing noticeable or strange in the reports..

**Comment by mitgr81** [comment]
> I don't know how nobody else is having this problem. Maybe they just don't send large messages

I wouldn't assume that.  😉 I've been silently watching and hoping you find more than I did when I looked.  We had some success changing servers from `daphne` to `uvicorn`.  We're still seeing some leakiness, but have resolved to using tools to monitor memory and restart services.

Here are some other things I've watched:
* https://github.com/ansible/awx/issues/7720
* https://github.com/django/channels/issues/1181
* https://github.com/yuriymironov96/django-channels-leak (from https://github.com/django/daphne/issues/373)

**Comment by cacosandon** [comment]
@mitgr81 what tools do you use to monitor and restart? For now I would love to implement that.

Will take a look on those resources!

**Comment by mitgr81** [comment]
> @mitgr81 what tools do you use to monitor and restart? For now I would love to implement that.

We're rocking a bespoke monitor for docker containers.  It's pretty simple; essentially we label each container with a valid restart time and a memory limit (among other rules); and the "container keeper" looks for them.

**Comment by mitgr81** [comment]
@cacosandon - Just curious if you've had any more luck than I have on this.

**Comment by cacosandon** [comment]
Hey @mitgr81! No updates yet.

We're using garbage collection on every message or new connection. This has helped a bit, but the memory still slowly increases and hits the max in about a week. We usually deploy and restart automatically the machines 2-3 times a week, which temporarily fixes the issue.

I hope the Channels team can look into this to see if it's a general problem with memory leaks. cc @carltongibson

**Comment by domvie** [comment]
We're running into the same issue. Daphne process used up over 50gigs of RAM on our server before it crashed.

**Comment by jetale** [comment]
> Hey @mitgr81! No updates yet.
> 
> We're using garbage collection on every message or new connection. This has helped a bit, but the memory still slowly increases and hits the max in about a week. We usually deploy and restart automatically the machines 2-3 times a week, which temporarily fixes the issue.
> 
> I hope the Channels team can look into this to see if it's a general problem with memory leaks. cc @carltongibson

Hi @cacosandon, Are you using uvicorn or daphne in production? or hypercorn?

**Comment by cacosandon** [comment]
> > Hey @mitgr81! No updates yet.
> 
> > 
> 
> > We're using garbage collection on every message or new connection. This has helped a bit, but the memory still slowly increases and hits the max in about a week. We usually deploy and restart automatically the machines 2-3 times a week, which temporarily fixes the issue.
> 
> > 
> 
> > I hope the Channels team can look into this to see if it's a general problem with memory leaks. cc @carltongibson
> 
> 
> 
> Hi @cacosandon, Are you using uvicorn or daphne in production? or hypercorn?

Hey! uvicorn for now.

**Comment by bigfootjon** [comment]
@cacosandon: What are the variables you haven't changed? It sounds like you've swapped everything out (including your application's business logic) and the problem still exists which is troubling.

Have you tried simplifying the code down until the problem doesn't exist? AIUI from this thread the channel-layer concept seems to be the cause, but have you tried to stub out the channel-layer code in various ways to see where the problem originates? (if it's not the channel-layer, then the same principle applies: just keep axing code until you've got the simplest program possible that still repros the problem)

(investing in a test harness that artificially generates problematic conditions might aid in discovering the problem by speeding up the testing cycle, if you haven't already done so)

**Comment by cacosandon** [comment]
Hey @bigfootjon!

Running this repository: https://github.com/cacosandon/django-channels-memory-leak, you'll notice the memory leaks.

If you remove the sending of large messages, then the problem disappears unless you open/close connections fast enough to make the memory go up again. It's literally the basic setup of Django Channels, so I don't know what else I should remove.

I think the next step is going deeper into Django Channels source code and start modifying things there. I don't have a lot of time now to do this, so we have mitigated it by monitoring and restarting our servers (for now).

**Comment by bigfootjon** [comment]
If you find some time to investigate I think removing code from channels is the right approach. If the memory charts aren’t doing it then the opposite (finding ways NOT to allocate memory) is the only path forward

**Comment by sevdog** [comment]
I tried to investigate this and in my tests I found an interesting hint: changing the "serializer" changes the memory footprint by a lot.

Since both `RedisChannelLayer` and `RedisChannelLayer` use msgpack I tried overriding the serialization in the first one with standardlib json and I got a very different memory profile:

With msgpack:
![image](https://github.com/user-attachments/assets/c059c3ea-fe74-4ffd-b851-646190ea958b)

```python
import json
import random
from channels_redis.core import RedisChannelLayer as _RedisChannelLayer


class RedisChannelLayer(_RedisChannelLayer):
    ### Serialization ###
    def serialize(self, message):
        """
        Serializes message to a byte string.
        """
        message = json.dumps(message).encode('utf-8')
        if self.crypter:
            message = self.crypter.encrypt(message)

        # As we use an sorted set to expire messages we need to guarantee uniqueness, with 12 bytes.
        random_prefix = random.getrandbits(8 * 12).to_bytes(12, "big")
        return random_prefix + message

    def deserialize(self, message):
        """
        Deserializes from a byte string.
        """
        # Removes the random prefix
        message = message[12:]

        if self.crypter:
            message = self.crypter.decrypt(message, self.expiry + 10)
        return json.loads(message.decode('utf-8'))

```


With JSON:
![image](https://github.com/user-attachments/assets/09a84dfe-eed9-4f4f-9b70-2cf5d8e2f373)

As you can see the memory in the JSON test return back to a "normal" level (there is still some memory which was not released, but much less than with msgpack).

I tested this on python 3.10 inside an alpine-docker container.

Also seems that there is a problem with msgpack and python 3.12: https://github.com/msgpack/msgpack-python/issues/612

I would like to have more time to perform furhter tests and learn how to better use the memory profile, for the moment I hope that this may help someoneelse to find a solution.

**Comment by acu192** [comment]
Cross linking https://github.com/django/channels/pull/1948 which is a long-standing known memory leak in channels.

**Comment by carltongibson** [comment]
Thanks @acu192. You're absolutely right. There's an unresolved chain of thought, and a likely fix sat there (for life reasons I suppose.)

@cacosandon if you could test the linked PR and feedback, that would help greatly.

**Comment by cacosandon** [comment]
Hey!

Yes, would love to help. Tried and raised some errors, surely because it's outdated and needs a rebase. Will ping him in the PR!

**Comment by carltongibson** [comment]
@cacosandon do note there's two related PRs. One for channels and one for channels-redis. You'll need to apply them both.

**Comment by cacosandon** [comment]
@carltongibson Already rebased both `channels` and `channels-redis` PRs. These are the updated dependencies:

```
Django==5.0.2
channels @ git+https://github.com/fosterseth/channels.git@clean_channels
channels-redis @ git+https://github.com/fosterseth/channels_redis.git@clean_channels
uvicorn[standard]==0.20.0
memory-profiler==0.61.0
memray==1.12.0
```

and these are the results for Uvicorn with PubSub:

<img width="1139" alt="pubsub-uvicorn" src="https://github.com/user-attachments/assets/bfbe94fa-b526-49f1-9dea-cfbb10cee3dc">


It seems the problem persists. I believe @sevdog's investigation around the serializer is likely the root cause, given its generic nature (whether using PubSub or not, and regardless of uvicorn, daphne, or hypercorn, even with a minimal example).

I can test with other settings later. Let me know!

**Comment by carltongibson** [comment]
@cacosandon OK, thanks for trying it. 

As to root cause, I still need to get a **minimal** reproduce nailed down here, but yes maybe... 

We're getting closer to it I suppose 😅

**Comment by bigfootjon** [comment]
@cacosandon, if you have some spare cycles could you try switching to the json serializer that @sevdog whipped up? https://github.com/django/channels_redis/pull/398

**Comment by cacosandon** [comment]
hey!

this is the test I'm running:
- send two messages, inside two different 'rooms'
- send 500 heavy messages like this 👇

```python
    async def receive(self, text_data):
        content = await self.decode_json(text_data)

        # Send message to room group
        await self.channel_layer.group_send(
            self.room_group_name, {"type": "chat.message", "content": content}
        )

        for i in range(500):
            # Create a struct of variable Mb from 1 to 5
            struct = bytearray(1024 * 1024 * random.randint(1, 5))
            message = bytes(struct)
            await self.channel_layer.group_send(
                self.room_group_name, {"type": "chat.binary", "message": message}
            )

            print(f"Sent message {i + 1} of 500")

            del struct
            gc.collect()
```

here is the repo: https://github.com/cacosandon/django-channels-memory-leak/blob/main/chat/consumers.py
here is a video explaining how to perform the test:

https://github.com/user-attachments/assets/8fd5d5db-dffd-4d46-b540-91324beb8659



here are the results:

## with original `requirements.txt`:
### uvicorn + redis channels
<img width="1139" alt="uvicorn-redis-old" src="https://github.com/user-attachments/assets/9a42af57-aa8c-42e9-aa5c-52e7ae7c476e">

### uvicorn + pubsub redis channels
<img width="1137" alt="uvicorn-pubsub-redis-old" src="https://github.com/user-attachments/assets/0e367acd-4af4-4edd-b9f8-82b8cf4c0aed">

## with https://github.com/django/channels_redis/pull/398
### uvicorn + redis channels
<img width="1141" alt="uvicorn-redis-new" src="https://github.com/user-attachments/assets/ef24522f-f783-4d48-9042-be5077ca9473">

### uvicorn + pubsub redis channels
<img width="1141" alt="uvicorn-pubsub-redis-new" src="https://github.com/user-attachments/assets/dc76c2c4-2e5e-44bc-a52c-380c4e639fd6">


seems like something improved! crazy how the memory went on a rollercoaster in the last one. 😅

**Comment by dogrocker** [comment]
> hey!
> 
> this is the test I'm running:
> 
> * send two messages, inside two different 'rooms'
> * on every message receive, send another 500 heavy messages like this 👇
> 
> ```python
>     async def receive(self, text_data):
>         content = await self.decode_json(text_data)
> 
>         # Send message to room group
>         await self.channel_layer.group_send(
>             self.room_group_name, {"type": "chat.message", "content": content}
>         )
> 
>         for i in range(500):
>             # Create a struct of variable Mb from 1 to 5
>             struct = bytearray(1024 * 1024 * random.randint(1, 5))
>             message = bytes(struct)
>             await self.channel_layer.group_send(
>                 self.room_group_name, {"type": "chat.binary", "message": message}
>             )
> 
>             print(f"Sent message {i + 1} of 500")
> 
>             del struct
>             gc.collect()
> ```
> 
> here is the repo: https://github.com/cacosandon/django-channels-memory-leak/blob/main/chat/consumers.py here is a video explaining how to perform the test (see in 2x):
> 
>  Screen.Recording.2024-09-22.at.12.59.41.1.mov 
> here are the results:
> 
> ## with original `requirements.txt`:
> ### uvicorn + redis channels
> <img alt="uvicorn-redis-old" width="1139" src="https://private-user-images.githubusercontent.com/30879716/369701152-9a42af57-aa8c-42e9-aa5c-52e7ae7c476e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NDAzMTgsIm5iZiI6MTczMDc0MDAxOCwicGF0aCI6Ii8zMDg3OTcxNi8zNjk3MDExNTItOWE0MmFmNTctYWE4Yy00MmU5LWFhNWMtNTJlN2FlN2M0NzZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDE3MDY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiMTU4MTViNWY4ZGMxYWM5MTQyNzFhNjgzNjExNzMzNWQyMjA2MTQxNGYxN2Y3NTFlMzAyNTIzMzNjNjBhY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rWxGBnuGySyjDTs-pPN9PsgHDEieLNANX4sKYb_L6p4">
> ### uvicorn + pubsub redis channels
> <img alt="uvicorn-pubsub-redis-old" width="1137" src="https://private-user-images.githubusercontent.com/30879716/369701163-0e367acd-4af4-4edd-b9f8-82b8cf4c0aed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NDAzMTgsIm5iZiI6MTczMDc0MDAxOCwicGF0aCI6Ii8zMDg3OTcxNi8zNjk3MDExNjMtMGUzNjdhY2QtNGFmNC00ZWRkLWI5ZjgtODJiOGNmNGMwYWVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDE3MDY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5MDlmMzNhOWQzODIyYTNlOWZlMDVjMTFkOWZhOWRhNGFiMmJjZGE4OGI1NDNkZmYzNzU1M2JjNzc1MmY5MDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OYPM1mTV2XklosXk9xvgK1CO49Kpy8OaLpTQQvzVAK0">
> ## with [django/channels_redis#398](https://github.com/django/channels_redis/pull/398)
> ### uvicorn + redis channels
> <img alt="uvicorn-redis-new" width="1141" src="https://private-user-images.githubusercontent.com/30879716/369701179-ef24522f-f783-4d48-9042-be5077ca9473.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NDAzMTgsIm5iZiI6MTczMDc0MDAxOCwicGF0aCI6Ii8zMDg3OTcxNi8zNjk3MDExNzktZWYyNDUyMmYtZjc4My00ZDQ4LTkwNDItYmU1MDc3Y2E5NDczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDE3MDY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwMzllNDgwNjYxYTEzNmJkMTU4NmQwNTIxZjNlNzc2ZDVkMjM1OGQ2YTQ0OGVhYmM2Y2M5YjM1YjNlOWNhNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ivy7Xoudg_rg2OFTSJyhICGJjAcoyXXAY-igqjlP2PA">
> ### uvicorn + pubsub redis channels
> <img alt="uvicorn-pubsub-redis-new" width="1141" src="https://private-user-images.githubusercontent.com/30879716/369701191-dc76c2c4-2e5e-44bc-a52c-380c4e639fd6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NDAzMTgsIm5iZiI6MTczMDc0MDAxOCwicGF0aCI6Ii8zMDg3OTcxNi8zNjk3MDExOTEtZGM3NmMyYzQtMmU1ZS00NGJjLWE1MmMtMzgwYzRlNjM5ZmQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDE3MDY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2Yjg1Y2QzZmE2ZDhjNmFhZTI4Zjk1YzE5NGMyYjU5ZTZhM2RkYjE2MDBlMDUwMjE5OGNjZTJiYTcwZGJmNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2gNtKh-h_is2hP5ZhgpYo2gt1w6m0je_pAoEM58f9AA">
> seems like something improved! crazy how the memory went on a rollercoaster in the last one. 😅

Hello! Do you mean use "pubsub redis" is fix the issue?

**Comment by ceelian** [comment]
Hi, is this happening in a rare edge case only or does that mean that every project that uses django channels will end up with a memory leak?

I wanted to use channels for a new project but now I am unsure if that issue will apply to my project as well. Thx.

**Comment by carltongibson** [comment]
@ceelian we're still waiting for someone to fully diagnose it. The trouble with these things is they're very set up specific. 

It seems to be an issue for folks pushing significant amounts of data through the channel layer. If you're using the channel layer for simple messaging it shouldn't be an issue. (If you have large data packages, you can always put them in a blob storage for collection, saving the channel layer just for notifications.) 

Plenty of folks are Channels successfully.

**Comment by ceelian** [comment]
@carltongibson thank you very much for the quick response and the explanation.

**Comment by lesc6** [comment]
even if you have small messages you will face this problem with memory, it just make it more slower to raise limits

**Comment by ceelian** [comment]
@lesc6 that's concerning, how do you/should we manage this in production systems?

**Comment by aDramaQueen** [comment]
As a workaround, if you run the ASGI server as SystemD unit (as I do), you may use following configuration:

```
# @see https://manpages.debian.org/testing/systemd/systemd.unit.5.en.html
# ATTENTION: Replace "<my-user-name>" with your actual user. Also replace "<redis/redict>" with either redis or redict, depending what your in memory database is.

[Unit]
Description=ASGI server (Hypercorn) for <My Project>
After=network.target
# (Optional) Start after the WSGI server
# Requires=<my-django-project>-wsgi.service
# After=<my-django-project>-wsgi.service

[Service]
Type=simple
User=<my-user-name>
Group=www-data
SupplementaryGroups=<redis/redict>
WorkingDirectory=/home/<my-user-name>/<my-django-project>
# (Optional) If you want to preload certain environment variables
# EnvironmentFile=/etc/path/to/environemtn.env
ExecStart=/home/<my-user-name>/<my-django-project>/.venv/bin/hypercorn --config file:/home/<my-user-name>/<my-django-project>/hypercorn.py asgi:core.asgi:application

# Restart behavior
Restart=always
RestartSec=3s
TimeoutStopSec=5

# This is a workaround for the memory leakage in Django Channels (@see https://github.com/django/channels/issues/2094)
# Ensure memory cgroup accounting is active (SystemD starts watching the memory consumption)
MemoryAccounting=yes
# Soft threshold - throttling/pressure starts here
MemoryHigh=400M
# Hard limit - kernel "Out Of Memory" (OOM) invoked for this cgroup
MemoryMax=512M
# Make kernel gracefully stop the whole cgroup if OOM hits (Choices: continue, stop, kill).
OOMPolicy=stop
# Make this service more likely to be chosen to be restarted (basically choose a number near to 1000)
OOMScoreAdjust=500

[Install]
WantedBy=multi-user.target
```

With this, you basically move the problem to SystemD. If you run out of the predefined Memory size, SystemD does stop the process and restarts it after the predifned number of seconds.

You may play around with the `MemoryHigh` & `MemoryMax` values and adjust this to your use-case.

It's not pretty, but better than a memory leakage...

**Comment by SwaY2000** [comment]
Hello! When it was fixed? More than 2 years we wait on fix
